Frameworks/Foundation/_NSInvocation.x86.mm

```diff
@@ -0,0 +1,238 @@
+            _returnLength = std::max(sizeof(uintptr_t), _returnLength);
+        }
+
+        auto nArguments = [_methodSignature numberOfArguments];
+        for (int i = 0; i < nArguments; ++i) {
+            _allocationExtents[i] = std::move(_allocateArgument([_methodSignature getArgumentTypeAtIndex:i]));
+        }
+
+        _buffer = new uint8_t[_offset];
+    };
+
+    ~impl() {
+        delete[] buffer;
+    }
+
+    allocation_extent _allocateArgument(const char* objcTypeEncoding) {
+        size_t nWords = std::max(1U, objc_aligned_size(objcTypeEncoding) / sizeof(uintptr_t));
+        auto length = nWords * sizeof(uintptr_t);
+
+        allocation_extent extent{ _offset, length };
+        _offset += length;
+        return extent;
+    }
+
+    void storeArgument(const void* value, unsigned int index) {
+        auto& extent = _allocationExtents[index];
+        const char* type = [_methodSignature getArgumentTypeAtIndex:index];
+        size_t size = objc_sizeof_type(type);
+
+        union {
+            uintptr_t u;
+            intptr_t i;
+        } itou; // "i" to "u"
+
+        { // integral arguments less than 4 bytes in width need sign- or zero-extension.
+            switch (type[0]) {
+                case _C_CHR:
+                    itou.i = *(signed char*)value;
+                    value = &itou.i;
+                    size = sizeof(intptr_t);
+                    break;
+                case _C_SHT:
+                    itou.i = *(signed short*)value;
+                    value = &itou.i;
+                    size = sizeof(intptr_t);
+                    break;
+                case _C_UCHR:
+                    itou.u = *(unsigned char*)value;
+                    value = &itou.u;
+                    size = sizeof(uintptr_t);
+                    break;
+                case _C_USHT:
+                    itou.u = *(unsigned short*)value;
+                    value = &itou.u;
+                    size = sizeof(uintptr_t);
+                    break;
+                default:
+                    // leave value/size alone; no extension required.
+                    break;
+            }
+        }
+
+        memcpy(_buffer + extent.offset, value, size);
+    }
+
+    void loadArgument(void* value, unsigned int index) const {
+        auto& extent = _allocationExtents[index];
+        const char* type = [_methodSignature getArgumentTypeAtIndex:index];
+        size_t size = objc_sizeof_type(type);
+        memcpy(value, _buffer + extent.offset, size);
+    }
+
+    StrongId<NSMethodSignature> _methodSignature;
+
+    uint8_t* _buffer;
+    off_t _offset;
+    size_t _returnLength;
+    return_type _returnType;
+
+    bool _stret;
+    allocation_extent _stretExtent;
+    std::vector<allocation_extent> _allocationExtents;
+};
+
+_NSInvocationCallFrame::_NSInvocationCallFrame(NSMethodSignature* methodSignature)
+    : _impl(new _NSInvocationCallFrame::impl(methodSignature)) {
```

Comment:
It probably makes sense to consider moving the struct definition to platform specific header and using the impl struct as a member (instead of heap allocated).  This will help with perf/fragmentation.